File: integration/testdata/go/httpingress/echo.go
//ftl:module echo
package echo

import (
	"context"
	"fmt"

	"ftl/builtin"

	"github.com/TBD54566975/ftl/go-runtime/ftl" // Import the FTL SDK.
)

type GetRequest struct {
	UserID string `json:"userId,omitempty"`
	PostID string `json:"postId,something,else"`
}

type Nested struct {
	GoodStuff string `json:"good_stuff"`
}

type GetResponse struct {
	Message string `json:"msg"`
	Nested  Nested `json:"nested"`
}

//ftl:verb
//ftl:ingress http GET /users/{userId}/posts/{postId}
func Get(ctx context.Context, req builtin.HttpRequest[GetRequest]) (builtin.HttpResponse[GetResponse, string], error) {
	return builtin.HttpResponse[GetResponse, string]{
		Headers: map[string][]string{"Get": {"Header from FTL"}},
		Body: ftl.Some(GetResponse{
			Message: fmt.Sprintf("UserID: %s, PostID: %s", req.Body.UserID, req.Body.PostID),
			Nested: Nested{
				GoodStuff: "This is good stuff",
			},
		}),
	}, nil
}

type PostRequest struct {
	UserID int `json:"user_id"`
	PostID int
}

type PostResponse struct {
	Success bool `json:"success"`
}

//ftl:verb
//ftl:ingress http POST /users
func Post(ctx context.Context, req builtin.HttpRequest[PostRequest]) (builtin.HttpResponse[PostResponse, string], error) {
	return builtin.HttpResponse[PostResponse, string]{
		Status:  201,
		Headers: map[string][]string{"Post": {"Header from FTL"}},
		Body:    ftl.Some(PostResponse{Success: true}),
	}, nil
}

type PutRequest struct {
	UserID string `json:"userId"`
	PostID string `json:"postId"`
}

type PutResponse struct{}

//ftl:verb
//ftl:ingress http PUT /users/{userId}
func Put(ctx context.Context, req builtin.HttpRequest[PutRequest]) (builtin.HttpResponse[builtin.Empty, string], error) {
	return builtin.HttpResponse[builtin.Empty, string]{
		Headers: map[string][]string{"Put": {"Header from FTL"}},
		Body:    ftl.Some(builtin.Empty{}),
	}, nil
}

type DeleteRequest struct {
	UserID string `json:"userId"`
}

type DeleteResponse struct{}

//ftl:verb
//ftl:ingress http DELETE /users/{userId}
func Delete(ctx context.Context, req builtin.HttpRequest[DeleteRequest]) (builtin.HttpResponse[builtin.Empty, string], error) {
	return builtin.HttpResponse[builtin.Empty, string]{
		Status:  200,
		Headers: map[string][]string{"Delete": {"Header from FTL"}},
		Body:    ftl.Some(builtin.Empty{}),
	}, nil
}

type HtmlRequest struct{}

//ftl:verb
//ftl:ingress http GET /html
func Html(ctx context.Context, req builtin.HttpRequest[HtmlRequest]) (builtin.HttpResponse[string, string], error) {
	return builtin.HttpResponse[string, string]{
		Headers: map[string][]string{"Content-Type": {"text/html; charset=utf-8"}},
		Body:    ftl.Some("<html><body><h1>HTML Page From FTL 🚀!</h1></body></html>"),
	}, nil
}

//ftl:verb
//ftl:ingress http POST /bytes
func Bytes(ctx context.Context, req builtin.HttpRequest[[]byte]) (builtin.HttpResponse[[]byte, string], error) {
	return builtin.HttpResponse[[]byte, string]{Body: ftl.Some(req.Body)}, nil
}

//ftl:verb
//ftl:ingress http GET /empty
func Empty(ctx context.Context, req builtin.HttpRequest[ftl.Unit]) (builtin.HttpResponse[ftl.Unit, string], error) {
	return builtin.HttpResponse[ftl.Unit, string]{Body: ftl.Some(ftl.Unit{})}, nil
}

//ftl:verb
//ftl:ingress http GET /string
func String(ctx context.Context, req builtin.HttpRequest[string]) (builtin.HttpResponse[string, string], error) {
	return builtin.HttpResponse[string, string]{Body: ftl.Some(req.Body)}, nil
}

//ftl:verb
//ftl:ingress http GET /int
func Int(ctx context.Context, req builtin.HttpRequest[int]) (builtin.HttpResponse[int, string], error) {
	return builtin.HttpResponse[int, string]{Body: ftl.Some(req.Body)}, nil
}

//ftl:verb
//ftl:ingress http GET /float
func Float(ctx context.Context, req builtin.HttpRequest[float64]) (builtin.HttpResponse[float64, string], error) {
	return builtin.HttpResponse[float64, string]{Body: ftl.Some(req.Body)}, nil
}

//ftl:verb
//ftl:ingress http GET /bool
func Bool(ctx context.Context, req builtin.HttpRequest[bool]) (builtin.HttpResponse[bool, string], error) {
	return builtin.HttpResponse[bool, string]{Body: ftl.Some(req.Body)}, nil
}

//ftl:verb
//ftl:ingress http GET /error
func Error(ctx context.Context, req builtin.HttpRequest[ftl.Unit]) (builtin.HttpResponse[ftl.Unit, string], error) {
	return builtin.HttpResponse[ftl.Unit, string]{
		Status: 500,
		Error:  ftl.Some("Error from FTL"),
	}, nil
}

//ftl:verb
//ftl:ingress http GET /array/string
func ArrayString(ctx context.Context, req builtin.HttpRequest[[]string]) (builtin.HttpResponse[[]string, string], error) {
	return builtin.HttpResponse[[]string, string]{
		Body: ftl.Some(req.Body),
	}, nil
}

type ArrayType struct {
	Item string `json:"item"`
}

//ftl:verb
//ftl:ingress http POST /array/data
func ArrayData(ctx context.Context, req builtin.HttpRequest[[]ArrayType]) (builtin.HttpResponse[[]ArrayType, string], error) {
	return builtin.HttpResponse[[]ArrayType, string]{
		Body: ftl.Some(req.Body),
	}, nil
}

File: integration/testdata/kotlin/httpingress/Echo.kt
package ftl.echo

import ftl.builtin.Empty
import ftl.builtin.HttpRequest
import ftl.builtin.HttpResponse
import kotlin.String
import kotlin.Unit
import xyz.block.ftl.Json
import xyz.block.ftl.Context
import xyz.block.ftl.HttpIngress
import xyz.block.ftl.Method
import xyz.block.ftl.Export

data class GetRequest(
  @Json("userId") val userID: String,
  @Json("postId") val postID: String,
)

data class Nested(
  @Json("good_stuff") val goodStuff: String,
)

data class GetResponse(
  @Json("msg") val message: String,
  @Json("nested") val nested: Nested,
)

data class PostRequest(
  @Json("user_id") val userId: Int,
  val postId: Int,
)

data class PostResponse(
  @Json("success") val success: Boolean,
)

data class PutRequest(
  @Json("userId") val userID: String,
  @Json("postId") val postID: String,
)

data class DeleteRequest(
  @Json("userId") val userID: String,
)

data class ArrayType(
  @Json("item") val item: String,
)

@Export
@HttpIngress(Method.GET, "/users/{userID}/posts/{postID}")
fun `get`(context: Context, req: HttpRequest<GetRequest>): HttpResponse<GetResponse, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Get" to arrayListOf("Header from FTL")),
    body = GetResponse(
      message = "UserID: ${req.body.userID}, PostID: ${req.body.postID}",
      nested = Nested(goodStuff = "This is good stuff")
    )
  )
}

@Export
@HttpIngress(Method.POST, "/users")
fun post(context: Context, req: HttpRequest<PostRequest>): HttpResponse<PostResponse, String> {
  return HttpResponse(
    status = 201,
    headers = mapOf("Post" to arrayListOf("Header from FTL")),
    body = PostResponse(success = true)
  )
}

@Export
@HttpIngress(Method.PUT, "/users/{userId}")
fun put(context: Context, req: HttpRequest<PutRequest>): HttpResponse<Empty, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Put" to arrayListOf("Header from FTL")),
    body = Empty()
  )
}

@Export
@HttpIngress(Method.DELETE, "/users/{userId}")
fun delete(context: Context, req: HttpRequest<DeleteRequest>): HttpResponse<Empty, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Delete" to arrayListOf("Header from FTL")),
    body = Empty()
  )
}

@Export
@HttpIngress(Method.GET, "/html")
fun html(context: Context, req: HttpRequest<Empty>): HttpResponse<String, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Content-Type" to arrayListOf("text/html; charset=utf-8")),
    body = "<html><body><h1>HTML Page From FTL 🚀!</h1></body></html>",
  )
}

@Export
@HttpIngress(Method.POST, "/bytes")
fun bytes(context: Context, req: HttpRequest<ByteArray>): HttpResponse<ByteArray, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Content-Type" to arrayListOf("application/octet-stream")),
    body = req.body,
  )
}

@Export
@HttpIngress(Method.GET, "/empty")
fun empty(context: Context, req: HttpRequest<Unit>): HttpResponse<Unit, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Empty" to arrayListOf("Header from FTL")),
    body = Unit
  )
}

@Export
@HttpIngress(Method.GET, "/string")
fun string(context: Context, req: HttpRequest<String>): HttpResponse<String, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("String" to arrayListOf("Header from FTL")),
    body = req.body
  )
}

@Export
@HttpIngress(Method.GET, "/int")
fun int(context: Context, req: HttpRequest<Int>): HttpResponse<Int, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Int" to arrayListOf("Header from FTL")),
    body = req.body
  )
}

@Export
@HttpIngress(Method.GET, "/float")
fun float(context: Context, req: HttpRequest<Double>): HttpResponse<Double, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Float" to arrayListOf("Header from FTL")),
    body = req.body
  )
}

@Export
@HttpIngress(Method.GET, "/bool")
fun bool(context: Context, req: HttpRequest<Boolean>): HttpResponse<Boolean, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("Bool" to arrayListOf("Header from FTL")),
    body = req.body
  )
}

@Export
@HttpIngress(Method.GET, "/error")
fun error(context: Context, req: HttpRequest<Unit>): HttpResponse<Boolean, String> {
  return HttpResponse(
    status = 500,
    headers = mapOf("Error" to arrayListOf("Header from FTL")),
    error = "Error from FTL"
  )
}

@Export
@HttpIngress(Method.GET, "/array/string")
fun arrayString(context: Context, req: HttpRequest<List<String>>): HttpResponse<List<String>, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("ArrayString" to arrayListOf("Header from FTL")),
    body = req.body
  )
}

@Export
@HttpIngress(Method.POST, "/array/data")
fun arrayData(context: Context, req: HttpRequest<List<ArrayType>>): HttpResponse<List<ArrayType>, String> {
  return HttpResponse(
    status = 200,
    headers = mapOf("ArrayData" to arrayListOf("Header from FTL")),
    body = req.body
  )
}
